### PR TITLE
cipher + improved hash

### LINKED ISSUE #460: encryption support
add an encryption protocol and droplet extension using CryptoEssentials.

### PR DESCRIPTION
**Pull request by tannernelson**
Adds encryption/decryption + improved hashing and crypto configuration.

Fixes #460 

Relies on https://github.com/vapor/crypto/pull/11

### Encryption

Encryption works like so:

```swift
let e = try drop.cipher.encrypt("vapor")
print(e) // "b7cMwL66ysKz7+vmKxoJLg=="
let d = try drop.cipher.decrypt(e)
print(d) // "vapor"
```

The configuration looks like this:

`crypto.json`
```json
{
    "cipher": {
        "method": "chacha20",
        "key": "passwordpasswordpasswordpassword",
        "iv": "password"
    }
} 
```

### Hash

The public API for hash looks the same, but it is now configurable with json.

`crypto.json`
```json
{
    "hash": {
        "method": "ripemd160",
        "key": "passwordpasswordpasswordpassword",
    }
} 
```

### CODE REVIEW

File: Sources/Vapor/Cipher/CipherProtocol.swift [hunk 1]
```diff
@@ -0,0 +1,97 @@
+public protocol CipherProtocol {
+    var defaultKey: Bytes { get }
+    var defaultIV: Bytes? { get }
+    func encrypt(_ bytes: Bytes, key: Bytes, iv: Bytes?) throws -> Bytes
+    func decrypt(_ bytes: Bytes, key: Bytes, iv: Bytes?) throws -> Bytes
+}
+
+import Cipher
```
Comment (LoganWright) [inline Sources/Vapor/Cipher/CipherProtocol.swift]: Let's move imports to top of file.

File: Sources/Vapor/Cipher/CipherProtocol.swift [hunk 2]
```diff
@@ -0,0 +1,97 @@
+public protocol CipherProtocol {
+    var defaultKey: Bytes { get }
+    var defaultIV: Bytes? { get }
+    func encrypt(_ bytes: Bytes, key: Bytes, iv: Bytes?) throws -> Bytes
```
Comment (LoganWright) [inline Sources/Vapor/Cipher/CipherProtocol.swift]: I'm not sure what `iv` is in this context, can we make a note for the less cryptographically minded

File: Sources/Vapor/Cipher/CipherProtocol.swift [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+    public func decrypt(_ string: String, key: Bytes? = nil, encoding: CipherEncoding = .base64) throws -> String {
+        let bytes: Bytes
+
+        switch encoding {
+        case .hex:
+            var data = Bytes()
+
+            var gen = string.characters.makeIterator()
+            while let c1 = gen.next(), let c2 = gen.next() {
```
Comment (LoganWright) [inline Sources/Vapor/Cipher/CipherProtocol.swift]: Can we use some more explicit variables here, I think it'll be easier to come back to this later

File: Sources/Vapor/Droplet/Droplet.swift [hunk 4]
```diff
@@ -148,15 +157,16 @@ public class Droplet {
...
+        cipher: CipherProtocol? = nil,
         console: ConsoleProtocol? = nil,
         log: Log? = nil,
         view: ViewRenderer? = nil,
         client: ClientProtocol.Type? = nil,
         database: Database? = nil,
         cache: CacheProtocol? = nil,
 
-        // middleware
+        // middlewarefl
```
Comment (LoganWright) [inline Sources/Vapor/Droplet/Droplet.swift]: fl

File: Tests/VaporTests/HashTests.swift [hunk 5]
```diff
@@ -1,39 +1,53 @@
...
+        let string = "vapor"
+        let defaultExpected = "97ce9a45eaf0b1ceafc3bba00dfec047526386bbd69241e4a4f0c9fde7c638ea"
+        let defaultKey = "123"
+
+        //test drop facade
+        let config = Config([
+            "crypto": [
+                "hash": [
+                    "method": "sha256",
+                    "key": .string(defaultKey)
```
Comment (tannernelson) [inline Tests/VaporTests/HashTests.swift]: @LoganWright any idea why I have to use `.string()` here? I tried doing `Config(node: ` to no avail.

File: Sources/Vapor/Droplet/Droplet.swift [hunk 6]
```diff
@@ -315,12 +326,120 @@ public class Droplet {
...
+
+        if let c = cipher as? CryptoCipher {
+            switch c.method {
+            case .chacha20:
+                if cipherKey.count != 32 {
+                    log.error("Chacha20 cipher key must be 32 bytes.")
+                }
+                if cipherIV == nil {
+                    log.error("Chacha20 cipher requires an initialization vector (iv).")
+                } else if cipherIV?.count != 8 {
```
Comment (LoganWright) [inline Sources/Vapor/Droplet/Droplet.swift]: Might be worth combining these two

```swift
Chacha20 initialization vector (iv) must be 8 bytes
```

maybe add 'and not nil' at the end.

File: Sources/Vapor/Droplet/Droplet.swift [hunk 7]
```diff
@@ -315,12 +326,120 @@ public class Droplet {
...
+                    log.error("Chacha20 initialization vector (iv) must be 8 bytes.")
+                }
+            case .aes128:
+                if cipherKey.count != 16 {
+                    log.error("AES-128 cipher key must be 16 bytes.")
+                }
+            case .aes256:
+                if cipherKey.count != 16 {
```
Comment (LoganWright) [inline Sources/Vapor/Droplet/Droplet.swift]: Sort of longer question, but is it safe to continue if these are invalid? Should we guard and throw maybe.

File: Sources/Vapor/Droplet/Droplet.swift [hunk 8]
```diff
@@ -315,12 +326,120 @@ public class Droplet {
...
         // from the configuration files or nothing.
-        let key = config["app", "key"]?.string
+        let hashKey = config["crypto", "hash", "key"]?.string?.bytes
         // initialize the hash from one provided
         // or use a default SHA2 hasher
-        let hash = provided.hash ?? SHA2Hasher(variant: .sha256, defaultKey: key)
+
+        let chosenHash: HashProtocol?
+
+        if let method = config["crypto", "cipher", "method"]?.string {
```
Comment (LoganWright) [inline Sources/Vapor/Droplet/Droplet.swift]: some of this stuff might be good on a config extension so we can start working towards a little bit leaner init

**Comment by LoganWright** [comment]
🚢 🇮🇹

File: Sources/Vapor/Droplet/Droplet.swift (hunk 7 above)
Comment (tannernelson) [inline Sources/Vapor/Droplet/Droplet.swift]: `drop.cipher.encrypt` will throw an error if called later on. So if user is never encrypting or decrypting I think it's fair to let them ignore.

**Pull request merged**